src/lib/provable/gadgets/blake2b.ts

```diff
@@ -0,0 +1,190 @@
+// https://www.blake2.net/blake2.pdf
+import { UInt64, UInt8 } from '../int.js';
+import { FlexibleBytes } from '../bytes.js';
+import { Bytes } from '../wrapped-classes.js';
+import { Gadgets } from './gadgets.js';
+
+export { BLAKE2B };
+
+const BLAKE2BConstants = {
+  IV: [
+    UInt64.from(0x6a09e667f3bcc908n),
+    UInt64.from(0xbb67ae8584caa73bn),
+    UInt64.from(0x3c6ef372fe94f82bn),
+    UInt64.from(0xa54ff53a5f1d36f1n),
+    UInt64.from(0x510e527fade682d1n),
+    UInt64.from(0x9b05688c2b3e6c1fn),
+    UInt64.from(0x1f83d9abfb41bd6bn),
+    UInt64.from(0x5be0cd19137e2179n),
+  ],
+
+  SIGMA: [
+    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
+    [14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3],
+    [11, 8, 12, 0, 5, 2, 15, 13, 10, 14, 3, 6, 7, 1, 9, 4],
+    [7, 9, 3, 1, 13, 12, 11, 14, 2, 6, 5, 10, 4, 0, 15, 8],
+    [9, 0, 5, 7, 2, 4, 10, 15, 14, 1, 11, 12, 6, 8, 3, 13],
+    [2, 12, 6, 10, 0, 11, 8, 3, 4, 13, 7, 5, 15, 14, 1, 9],
+    [12, 5, 1, 15, 14, 13, 4, 10, 0, 7, 6, 3, 9, 2, 8, 11],
+    [13, 11, 7, 14, 12, 1, 3, 9, 5, 0, 15, 4, 8, 6, 2, 10],
+    [6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5],
+    [10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0],
+  ],
+};
+
+const BLAKE2B = {
+  hash(data: FlexibleBytes, digestLength = 64) {
+    const state = initialize(digestLength);
+    update(state, Bytes.from(data).bytes);
+    const out = final(state);
+    return Bytes.from(out);
+  },
+  get IV() {
+    return BLAKE2BConstants.IV;
+  },
+};
+
+function initialize(outlen: number): {
+  h: UInt64[];
+  t: UInt64[];
+  f: UInt64[];
+  buf: UInt8[];
+  buflen: number;
+  outlen: number;
+} {
+  const h = BLAKE2B.IV.slice(); // shallow copy IV to h
+  h[0] = UInt64.from(0x01010000).xor(UInt64.from(outlen)).xor(h[0]);
```

Comment:
`UInt64.from(0x01010000)` puts a constant Field element inside `UInt64.value`
same for `UInt64.from(outlen)`, because `outLen` is a `number` just like `0x01010000` (the fact that it's a function argument rather than a constant makes no difference)

so here, `h[0]` still gets assigned a constant again!

when that constant is combined with variables in the circuit, it works just like in snarky. i.e. the constant is either folded into gate coefficients, or, most commonly, a dedicated generic gate is added that creates a variable fixed to be equal to this constant